Update AppendCollection schemas and algorithms (#91)
ceramicnetwork/CIP: CIPs/CIP-85/CIP-85.md

The `AppendCollection` schema must be an `object` with a `$ceramic` field from [CIP-88](../CIP-88/CIP-88.md) having the type `appendCollection` and pointing to the slice's `schema`, along with the following properties:

- `sliceMaxItems`: the maximum number of items a single slice should contain
- `slicesCount`: the total number of slices the collection contains

```js
{
  $schema: 'http://json-schema.org/draft-07/schema#',
  $ceramic: { type: 'appendCollection', sliceSchema: '<slice schema ID>' },
  title: 'MyCollection',
  type: 'object',
  properties: {
    sliceMaxItems: { type: 'integer', minimum: 10, maximum: 256 },

>     slicesCount: { type: 'integer', minimum: 1 },

Reviewer: ```suggestion
    slicesCount: { type: 'integer', minimum: 0 },
```

Reviewer: Needed to create the initial empty AppendCollection to get the DocID

Author: I'm wondering if really it's needed?
Setting it to `0` to get started means creating the collection, then the slice, and finally updating the collection, but if we set the `slicesCount` to `1` when creating the collection, we can then create the slice and no need to update the collection.

Reviewer: Okay yeah good point.  Worst case scenario when you load the index 0 slice it has no content yet - no biggie.

  },
  required: ['sliceMaxItems', 'slicesCount']
}
```

#### CollectionSlice schema

The `CollectionSlice` schema must be an `object` with a `$ceramic` field from [CIP-88](../CIP-88/CIP-88.md) having the type `collectionSlice` and pointing to the slice's `schema`, along with the following properties:

- `collection`: the DocID of the collection the slice is part of
- `sliceIndex`: index of the slice in the collection, between `0` and the collection's `slicesCount` minus `1`
- `contents`: array with a `maxItems` value matching the `AppendCollection` `sliceMaxItems` property and defining the items schemas, that must include `{ type: 'null' }` in order to support removals

```js
{
  $schema: 'http://json-schema.org/draft-07/schema#',
  $ceramic: { type: 'collectionSlice' },

>   title: 'MyCollectionSLice',

Reviewer: ```suggestion
  title: 'MyCollectionSlice',
```

  type: 'object',
  properties: {
    collection: { type: 'string', maxLength: 150 },
    sliceIndex: { type: 'integer', minimum: 0 },
    contents: {
      type: 'array',
      maxItems: 256,
      minItems: 0,
      items: {
        oneOf: [{ type: 'object', ... }, { type: 'null' }],
      },
    },
  },
  required: ['collection', 'sliceIndex', 'contents'],
}
```

### Algorithms

#### Deterministic slice access

Accessing slices in the collection relies on Ceramic's ability to load documents deterministically based on their genesis contents:

- `collection`: the collection's `DocID` string
- `sliceIndex`: the slice's index, an integer between `0` (first slice) and `slicesCount - 1` (last slice)
- `contents`: empty array

#### First insertion

> This flow assumes a prerequisite check that the `AppendCollection` document has not been created yet.

> 1. Create the `AppendCollection` document.

Reviewer: ```suggestion
1. Create the `AppendCollection` document with `slicesCount` of `0`
```

1. Create a deterministic `CollectionSlice` document with `sliceIndex` of `0` and an empty `contents` array.
1. Update the created `CollectionSlice` document with the `contents` array containing the item to insert.

> 1. Create the `AppendCollection` document with a `slicesCount` of `1`.

Reviewer: ```suggestion
1. Update the `AppendCollection` document with a `slicesCount` of `1`.
```

Reviewer: This step can then be removed entirely if the first step is to create the `AppendCollection` doc with a slicesCount of 1

  title: 'MyCollectionSLice',
  type: 'object',
  properties: {

>     collection: { type: 'string', maxLength: 150 },

Reviewer: Don't we have a new special way to refer to ceramic DocIDs in json-schemas?

Author: Yes we could add the `$ceramic` field here to be more explicit, but we already have `$ceramic: { type: 'collectionSlice', schema: '<slice schema ID>' }` on the object itself so I would think this is enough information for the tools to act on, knowing what properties are contained by the `collectionSlice`?

    sliceNumber: { type: 'integer', minimum: 0 },
    contents: {
      type: 'array',
      maxItems: 256,
      minItems: 0,
      items: {
        oneOf: [{ type: 'object', ... }, { type: 'null' }],
      },
    },
  },
  required: ['collection', 'sliceNumber', 'contents'],
}
```

### Algorithms

#### Deterministic slice access

Accessing slices in the collection relies on Ceramic's ability to load documents deterministically based on their genesis contents:

- `collection`: the collection's `DocID` string
- `sliceNumber`: the slice's number, an integer between `0` (first slice) and `slicesCount - 1` (last slice)
- `contents`: empty array

#### First insertion

> This flow assumes a prerequisite check that the `AppendCollection` document has not been created yet.

> 1. Create the `CollectionSlice` document with the `contents` array containing the item to insert.

Reviewer: This needs to first be created with empty `contents` array then updated to add the item

Author: Good catch, thanks! I updated it.

1. Create the `AppendCollection` document with a `slicesCount` of `1`.

#### Other insertions

1. Load the `AppendCollection` document.

> 1. Load the `CollectionSlice` document based on its determistic content, using the `slicesCount` from the collection.

Reviewer: ```suggestion
1. Load the most recent `CollectionSlice` document based on its deterministic content, using the `slicesCount` from the collection.
```

1. Check the length of the `contents` array of the `CollectionSlice`:

- If it is lower than the `sliceSize` value of the `AppendCollection`, add the item to the `contents` array.
- If it is equal to the `sliceSize` value:

>   1. Create a new `CollectionSlice` document with the `contents` array containing the item to insert.

Reviewer: again, needs to start with empty array then get updated

> This flow assumes a prerequisite check that the `AppendCollection` document has not been created yet.

1. Create a deterministic `CollectionSlice` document with `sliceNumber` of `0` and an empty `contents` array.
1. Update the created `CollectionSlice` document with the `contents` array containing the item to insert.

> 1. Create the `AppendCollection` document with a `slicesCount` of `1`.

Reviewer: hmm... I just realized it may be difficult to create the `CollectionSlice` before the `AppendCollection` since you need the DocID from the `AppendCollection` in order to make the `CollectionSlice`

Author: 👍 right, good catch! I updated to start the flow with the `AppendCollection` document creation
